File: src/bot/path_planner.cpp
#include <ros/ros.h>
#include <std_msgs/Float32MultiArray.h>
#include <vector>
#include <queue>

class PathPlanner
{
public:
    PathPlanner()
    {
        costmap_sub = nh.subscribe("costmap_topic", 1, &PathPlanner::updateCostmap, this);
        goal_sub = nh.subscribe("bot_goal_topic", 1, &PathPlanner::updateGoal, this);
        bot_sub = nh.subscribe("bot_loc_topic", 1, &PathPlanner::updateBot, this);
        path_pub = nh.advertise<std_msgs::Float32MultiArray>("path_topic", 10);
    }

    void updateCostmap(const std_msgs::Float32MultiArray::ConstPtr &data)
    {
        costmap = data->data;
    }

    void updateGoal(const std_msgs::Float32MultiArray::ConstPtr &data)
    {
        goal = data->data;
    }

    void updateBot(const std_msgs::Float32MultiArray::ConstPtr &data)
    {
        bot = data->data;
    }

    void pathPlanning(float botID)
    {
        for (auto &i : goal)
        {
            if (i[0] != botID)
            {
                costmap[i[1]][i[2]] = 1;
            }
        }

        std::vector<float> curLoc = getLoc(botID);
        std::vector<float> tarLoc = getLoc(botID);
        std_msgs::Float32MultiArray path = dijkstra(curLoc, tarLoc);
        path.push_front(botID);
        path_pub.publish(path);
    }

    void publishPath()
    {
        for (auto &i : bot)
        {
            pathPlanning(i[0]);
        }
    }

    std::vector<float> getLoc(float botID)
    {
        for (auto &item : bot)
        {
            if (item[0] == botID)
            {
                float x = item[1];
                float y = item[2];
                std::vector<float> curLoc = {x, y};
                return curLoc;
            }
        }
        return {};
    }

    std_msgs::Float32MultiArray dijkstra(const std::vector<float> &curLoc, const std::vector<float> &tarLoc)
    {
        int rows = costmap.size();
        int cols = costmap[0].size();
        std::vector<std::vector<float>> distances(rows, std::vector<float>(cols, std::numeric_limits<float>::infinity()));
        distances[curLoc[0]][curLoc[1]] = 0;
        std::vector<std::vector<std::pair<int, int>>> prev(rows, std::vector<std::pair<int, int>>(cols, {-1, -1}));
        std::priority_queue<std::pair<float, std::pair<int, int>>, std::vector<std::pair<float, std::pair<int, int>>>> pq;
        pq.push({0, {curLoc[0], curLoc[1]}});

        while (!pq.empty())
        {
            float dist = pq.top().first;
            int x = pq.top().second.first;
            int y = pq.top().second.second;
            pq.pop();
            if (x == tarLoc[0] && y == tarLoc[1])
            {
                break;
            }
            if (dist > distances[x][y])
            {
                continue;
            }

            std::vector<std::pair<int, int>> neighbors = getNeighbors(x, y, costmap);
            for (const auto &neighbor : neighbors)
            {
                int nx = neighbor.first;
                int ny = neighbor.second;
                float newDist = dist + 1; // Assuming each step has a distance of 1
                if (newDist < distances[nx][ny])
                {
                    distances[nx][ny] = newDist;
                    prev[nx][ny] = {x, y};
                    pq.push({newDist, {nx, ny}});
                }
            }
        }

        std::vector<float> path;
        int currX = tarLoc[0];
        int currY = tarLoc[1];
        while (currX != -1 && currY != -1)
        {
            path.push_back( currX);
            path.push_back( currY);
            int prevX = prev[currX][currY].first;
            int prevY = prev[currX][currY].second;
            currX = prevX;
            currY = prevY;
        }

        std_msgs::Float32MultiArray pathMsg;
        pathMsg.data = path;
        return pathMsg;
    }

    std::vector<std::vector<float>> getNeighbors(float x, float y)
    {
        int rows = costmap.size();
        int cols = costmap[0].size();
        std::vector<std::vector<float>> neighbors;

        if (x > 0 && costmap[static_cast<int>(x) - 1][static_cast<int>(y)] == 0)
        {
            neighbors.push_back({x - 1, y});
        }
        if (x < rows - 1 && costmap[static_cast<int>(x) + 1][static_cast<int>(y)] == 0)
        {
            neighbors.push_back({x + 1, y});
        }
        if (y > 0 && costmap[static_cast<int>(x)][static_cast<int>(y) - 1] == 0)
        {
            neighbors.push_back({x, y - 1});
        }
        if (y < cols - 1 && costmap[static_cast<int>(x)][static_cast<int>(y) + 1] == 0)
        {
            neighbors.push_back({x, y + 1});
        }

        return neighbors;
    }

private:
    ros::NodeHandle nh;
    ros::Subscriber costmap_sub;
    ros::Subscriber goal_sub;
    ros::Subscriber bot_sub;
    ros::Publisher path_pub;
    std::vector<float> costmap;
    std::vector<float> goal;
    std::vector<float> bot;
    int costmap_width = 14;  // Adjust these values according to your data
    int costmap_height = 14; // Adjust these values according to your data
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "path_planner");
    PathPlanner pathPlanner;
    pathPlanner.publishPath();
    ros::spin();
    return 0;
}

File: src/bot/scripts/path_planner.py
import rospy
from std_msgs.msg import Float32MultiArray
import heapq

class path_planner:
    def __init__(self):
        # 14x14 matrix
        self.costmap = Float32MultiArray() 

        # nx3 matrix [[botID,x,y],[botID,x,y],[botID,x,y]...]
        self.goal = Float32MultiArray() 

        # matrix [[botID,x1,y1,x2,y2,...],[botID,x1,y1,x2,y2,...],...]
        self.path = Float32MultiArray() 

        # nx3 matrix [[botID,x,y],[botID,x,y],[botID,x,y]...]
        self.bot = Float32MultiArray() 

        self.path_publisher = rospy.Publisher('path_topic', Float32MultiArray, queue_size=10)

        rospy.Subscriber('costmap_topic', Float32MultiArray, self.update_costmap)
        rospy.Subscriber('bot_goal_topic', Float32MultiArray, self.update_goal)
        rospy.Subscriber('bot_loc_topic', Float32MultiArray, self.update_bot)


    def update_costmap(self,data):
        self.costmap = data.data
    
    def update_goal(self,data):
        self.goal = data.data

    def update_bot(self,data):
        self.bot = data.data

    def path_planning(self,botID):
        
        for i in self.goal:
            if i[0] != botID:
                self.costmap[i[1],i[2]] = 1
        
        cur_loc = self.get_loc(botID)
        tar_loc = self.get_loc(botID)
        path  = self.dijkstra(cur_loc, tar_loc)
        self.path_publisher.publish(path)

    def publish_path(self):
        for i in self.bot:
            self.path_planning(i[0])

    def get_loc(self, bot_id):
        for item in self.bot:
            if item[0] == bot_id:
                x, y = item[1], item[2]
                cur_loc = [x, y]
                return cur_loc

    def dijkstra(self, cur_loc, tar_loc):
        rows = len(self.costmap)
        cols = len(self.costmap[0])
        distances = [[float('inf')] * cols for _ in range(rows)]
        distances[cur_loc[0]][cur_loc[1]] = 0
        visited = set()
        prev = [[None] * cols for _ in range(rows)]
        pq = [(0, cur_loc)]

        while pq:
            dist, node = heapq.heappop(pq)
            if node == tar_loc:
                break
            if node in visited:
                continue
            visited.add(node)

            neighbors = self.get_neighbors(node[0], node[1])
            for neighbor in neighbors:
                if neighbor in visited:
                    continue
                new_dist = dist + 1  # Assuming each step has a distance of 1
                if new_dist < distances[neighbor[0]][neighbor[1]]:
                    distances[neighbor[0]][neighbor[1]] = new_dist
                    prev[neighbor[0]][neighbor[1]] = node
                    heapq.heappush(pq, (new_dist, neighbor))

        path = self.reconstruct_path(prev, tar_loc)
        return path
   
                
    
def main():
    rospy.init_node('path_planner', anonymous=True)
    pp = path_planner()
    pp.publish_path()

    # Spin to keep the node running
    rospy.spin(10)

if __name__ == '__main__':
    main()